File: typescript/src/caching/cache.ts
export interface Cache {
  save(key: string, value: any): Promise<void>;
  load(key: string): Promise<any>;
  age(key: string): Promise<number>;
}

export const IMAGE_KEY = 'ImageKey';

interface CacheData {
  url: string;
  imageData: ArrayBuffer;
  creation: number;
}

export class ImageCache implements Cache {
  constructor(private kv: KVNamespace) {}

  async age(key: string): Promise<number> {
    const data = await this.getCacheData(key);
    if (!data) {
      throw new Error(`invalid key provided ${key}`);
    }

    return Date.now() - data.creation;
  }

  async save(key: string, value: any): Promise<void> {
    const url = value as string;
    if (typeof url !== 'string') {
      throw new Error('invalid value provided to cache manager');
    }

    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`failed to fetch image: ${response.status}`);
    }

    const imageData = await response.arrayBuffer();
    
    const cacheData: CacheData = {
      url,
      imageData,
      creation: Date.now(),
    };

    await this.kv.put(key, JSON.stringify({
      url,
      creation: cacheData.creation,
    }));
    
    await this.kv.put(`${key}_data`, imageData);
  }

  async load(key: string): Promise<ReadableStream> {
    const data = await this.getCacheData(key);
    if (!data) {
      throw new Error(`invalid key provided ${key}`);
    }

    const imageData = await this.kv.get(`${key}_data`, 'arrayBuffer');
    if (!imageData) {
      throw new Error(`image data not found for key ${key}`);
    }

    return new ReadableStream({
      start(controller) {
        controller.enqueue(new Uint8Array(imageData));
        controller.close();
      }
    });
  }

  private async getCacheData(key: string): Promise<{ url: string; creation: number } | null> {
    const data = await this.kv.get(key);
    if (!data) {
      return null;
    }

    try {
      return JSON.parse(data);
    } catch {
      return null;
    }
  }
}

export function newImageCache(kv: KVNamespace): Cache {
  return new ImageCache(kv);
}

File: go/caching/cache.go
package caching

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"time"
)

const (
	ImageKey = "ImageKey"
)

type Cache interface {
	Save(key string, value interface{}) error
	Load(key string) (interface{}, error)
	Age(key string) (time.Duration, error)
}

type cacheData struct {
	path     string
	creation time.Time
}

type ImageCache struct {
	cache map[string]cacheData
}

func NewImage() Cache {
	return ImageCache{cache: map[string]cacheData{}}
}

func (i ImageCache) Age(key string) (time.Duration, error) {
	if _, ok := i.cache[key]; !ok {
		return 0, fmt.Errorf("invalid key provided %v", key)
	}

	return time.Since(i.cache[key].creation), nil
}

func (i ImageCache) Save(key string, value interface{}) error {
	url, ok := value.(string)
	if !ok {
		return fmt.Errorf("invalid value provided to cache manager")
	}

	response, err := http.Get(url)
	if err != nil {
		return err
	}
	defer response.Body.Close()

	//open a file for writing
	file, err := os.Create("/tmp/image" + filepath.Ext(url))
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = io.Copy(file, response.Body)
	if err != nil {
		return err
	}

	i.cache[ImageKey] = cacheData{path: "/tmp/image" + filepath.Ext(url), creation: time.Now()}

	return nil
}

func (i ImageCache) Load(key string) (interface{}, error) {
	if _, ok := i.cache[key]; !ok {
		return nil, fmt.Errorf("invalid key provided %v", key)
	}

	return os.Open(i.cache[key].path)
}
